content-files: add alternate content backing store

garlick opened the pull request:
This PR adds a new content backing store option, mainly as a demo, with lots of inline comments.

For now it just stores objects in one big flat directory using blobrefs as the file names, which will not be very scalable (depending on the file system used).

To allow flux to be invoked with a non-default backing store module, one can start flux with e.g.
```
flux start -o,-Scontent.backing-module=content-files
```
This attribute is peeked at by the rc1 script.  This needs more thought and probably isn't the complete or final solution to #2984.

Still more to do here but wanted to get this up early for @garrettbslone to peek at when he's ready.

garlick commented:
Added tests and rebased on current master.

garlick commented:
This PR could benefit from some review now I think (dropping WIP)

To recap: this adds another content backing store, it factors some reusable code out of `content-sqlite` so it doesn't have to be repeated in every backing store, and it adds a mechanism to select the backing store on the broker command line.

One thing that's missing is a a TOML config hook to select the backing store, and a way to configure the module itself.  I would propose to do that in another PR.

chu11 commented on src/modules/content-files/content-files.c:
```diff
@@ -0,0 +1,345 @@
+        flux_log_error (h, "content.blob-size-limit");
+        goto error;
+    }
+    ctx->blob_size_limit = strtoul (tmp, NULL, 10);
```
should check for error on blob-size-limit input?  Noticed not done in `content-sqlite`.  perhaps test for invalid blob-size-limit good idea too?

chu11 commented on src/modules/content-files/content-files.c:
```diff
@@ -0,0 +1,345 @@
+     * - value is the db directory
+     * - if it exists, preserve existing content; else create empty
+     * Otherwise:
+     * - ${rundir}/content.files is the backing file
```
backing path instead of backing file?

chu11 commented on src/modules/content-files/filedb.h:
```diff
@@ -0,0 +1,43 @@
+ * and 0 is returned.  On failure, -1 is returned with errno set.
+ * Pass '*errstr' in pre-set to NULL and if a human readable error message
+ * is appropriate, it is assigned on error.
+ */
```
should clarify `datap` should be freed, but `errstr` not freed?

garlick replied:
Will do!

garlick commented:
Just pushed an update that I think addresses your comments @chu11.

Regarding the `content.blob-size-liimit` check - I went ahead and dropped the check from the backing store(s) and the backing store sharness tests.  It's checked in the cache, and guests don't have direct access to the backing stores (nor is the interface used anywhere except in testing), so I think it's probably not needed.

I rebased on current master to pick up the test reliability improvements, and squashed the incremental development.

garlick commented:
Fixed a sharness test && chain issue and forced a push.

garlick commented:
Pushed a few extra tests but looks like pylint is failing over here too, like in #2960... Hrm.

garlick commented:
Rebased on current master to pick up pylint fixes from #3035

garlick commented:
Restarted workflow after what looked like an unlucky timeout in the fluxion tests.

garlick commented:
Restarted a couple travis builders that seemed to hang in odd places.  In the one that got as far as sharness tests, the three tests modified by this PR had completed so I assume this is unrelated to the PR.

chu11 commented on src/modules/content-files/test/filedb.c:
```diff
@@ -0,0 +1,128 @@
+    errno = 0;
+    ok (filedb_get (dbpath, "noexist", &data, &size, &errstr) < 0
+        && errno == ENOENT,
+        "filedb_get key="" failed with ENOENT");
```
missing escape chars?

chu11 commented on src/modules/content-files/test/filedb.c:
```diff
@@ -0,0 +1,128 @@
+
+    errno = 0;
+    errstr = NULL;
+    ok (filedb_put (dbpath, "", "", 2, &errstr) < 0 && errno == EINVAL,
```
although it probably doesn't matter for this and the next test, but size should be 1?

chu11 commented:
@garlick overall looks good, just the two nits I noticed above.  I think you said on the call you needed to squash?  I'd say squash and we're good to go.

garlick commented:
OK, those are fixed.  I'll go ahead and squash.  Thanks!

chu11 commented:
I went ahead and approved, set MWP when you think it's ready.

garlick commented:
Thanks!
